### PR TITLE
Prometheus Metrics table

### PR DESCRIPTION
**Pull request by allanliu**
- Table consists of metrics scraped from configured target urls, where the timestamp column/field is a unix timestamp in MS taken when the response is received.
- In order to use the table, you must pass in a JSON config file consisting of target urls in this format
```bash
osqueryi --config_path ./example.json
```

JSON sample:
```json
{
  "prometheus_targets": {
    "urls": [
      "http://localhost:9100/metrics",
      "http://localhost:9101/metrics"
    ]
  }
}
```

### CODE REVIEW

File: osquery/tables/applications/posix/prometheus_metrics.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+};
+
+class PrometheusMetrics {
+ public:
+  PrometheusMetrics(std::vector<std::string> urls, int timeoutDurationS = 1)
+      : options_(http::client::options().follow_redirects(true).timeout(
+            timeoutDurationS)),
+        client_(options_) {
+    for (const auto& url : urls) {
+      retData* rd = new retData;
```
Comment (jedi22) [inline osquery/tables/applications/posix/prometheus_metrics.h]: Does this need to be on the heap? Can you just create this on the stack and copy it into scrapeResults_. Make sure you update all other references to no longer be a pointer. You can probably do this all without allocating memory on the heap.

File: osquery/tables/applications/posix/prometheus_metrics.h [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      scrapeResults_[url] = rd;
+    }
+  }
+
+  virtual ~PrometheusMetrics() {
+    for (auto& target : scrapeResults_) {
+      delete target.second;
```
Comment (jedi22) [inline osquery/tables/applications/posix/prometheus_metrics.h]: If you can move everything onto the stack you won't need this anymore.

File: osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+  validatePMTest(pm, expected);
+}
+
+TEST_F(PrometheusMetricsTest, happy_path_0_metrics) {
+  // Initialize stubbed scrape results.
+  std::chrono::milliseconds now(
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          std::chrono::system_clock::now().time_since_epoch()));
+
+  retData* r0 = new retData{"", now};
```
Comment (jedi22) [inline osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp]: Same as above, this can be almost certainly be done on the stack.

File: osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+
+  validatePMTest(pm, expected);
+}
+
+TEST_F(PrometheusMetricsTest, happy_path_1_metric) {
+  // Initialize stubbed scrape results.
+  std::chrono::milliseconds now(
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          std::chrono::system_clock::now().time_since_epoch()));
+  retData* r0 = new retData{
```
Comment (jedi22) [inline osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp]: Same

File: osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+}
+
+TEST_F(PrometheusMetricsTest, happy_path_10_metrics_1_target) {
+  // Initialize stubbed scrape results.
+  std::chrono::milliseconds now(
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          std::chrono::system_clock::now().time_since_epoch()));
+  std::string nowStr(std::to_string(now.count()));
+
+  retData* r0 = new retData{
```
Comment (jedi22) [inline osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp]: Even in tests, we shouldn't need new.

File: osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+      "gauge\nprocess_resident_memory_bytes 1.0170368e+07\n# HELP "
+      "process_start_time_seconds Start time of the process since unix epoch "
+      "in seconds.\n# TYPE process_start_time_seconds "
+      "gauge\nprocess_start_time_seconds 1.48475733855e+09\n# HELP "
+      "process_virtual_memory_bytes Virtual memory size in bytes.\n# TYPE "
+      "process_virtual_memory_bytes gauge\nprocess_virtual_memory_bytes "
+      "1.3934592e+07\n",
+      now,
+  };
+  retData* r1 = new retData{
```
Comment (jedi22) [inline osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp]: Same

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    std::string dest;
+    std::string ts(std::to_string(target.second->timestampMS.count()));
+
+    while (std::getline(ss, dest)) {
+      if (dest[0] != '#') {
+        std::stringstream iss(dest);
+        std::string idest;
+        std::vector<std::string> metric;
+
+        while (std::getline(iss, idest, ' ')) {
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: [We have a helper function](https://github.com/facebook/osquery/blob/master/osquery/core/conversions.h#L64) `osquery::split` that might better suite what you're trying to do here and would save on local variables. Then you could change your loop to look something like:
`for(const auto& tok : osquery::split(dest, ' '))`

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+  const auto& config = parser->getData().get_child(configParserRootKey);
+
+  if (config.count("urls") == 0) {
+    return result;
+  }
+
+  const auto& urls = config.get_child("urls");
+  std::vector<std::string> iurls;
+  for (const auto& url : urls) {
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: This seems redundant, why not just `for (const auto& urls : config.get_child("urls"))`?

File: osquery/tables/applications/posix/prometheus_metrics.h [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+const std::string col_timestamp = "timestamp_ms";
+
+struct retData {
+  std::string content;
+  std::chrono::milliseconds timestampMS;
+};
+
+class PrometheusMetrics {
+ public:
+  PrometheusMetrics(std::vector<std::string> urls, int timeoutDurationS = 1)
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.h]: If you're only using `urls` for `const` iteration, then make this parameter pass by const ref, `const std::vector<std::string>& urls`

Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.h]: nit: `timeoutDurationS`? Was this supposed to be `timeoutDurations`?

File: osquery/tables/applications/posix/prometheus_metrics.h (hunk 1 above)
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.h]: I think you're right, it shouldn't need to be on the heap.  Will change and fix.

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+namespace tables {
+
+void PrometheusMetrics::parseScrapeResults() {
+  for (auto const& target : scrapeResults_) {
+    std::stringstream ss(target.second->content);
+    std::string dest;
+    std::string ts(std::to_string(target.second->timestampMS.count()));
+
+    while (std::getline(ss, dest)) {
+      if (dest[0] != '#') {
```
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: Good call.  Will change.

File: osquery/tables/applications/posix/prometheus_metrics.cpp (hunk 7 above)
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: Cool.  I'll give it a shot.

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+        std::stringstream iss(dest);
+        std::string idest;
+        std::vector<std::string> metric;
+
+        while (std::getline(iss, idest, ' ')) {
+          if (dest != "") {
```
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: ok

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+      http::client::response response(client_.get(request));
+
+      target.second->timestampMS =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              std::chrono::system_clock::now().time_since_epoch());
+      target.second->content = static_cast<std::string>(body(response));
+
+    } catch (std::exception& e) {
+      LOG(ERROR) << "failed on scrape of target " << target.first
+                 << " with error: " << e.what();
```
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: ok

File: osquery/tables/applications/posix/prometheus_metrics.cpp (hunk 8 above)
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: Yep you're right, it's redundant.

File: osquery/tables/applications/posix/prometheus_metrics_utils.h [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+#include <string>
+#include <vector>
+
+#include <osquery/config.h>
+
+namespace pt = boost::property_tree;
+
+namespace osquery {
+
+const std::string configParserRootKey("prometheus_targets");
```
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics_utils.h]: ok

File: osquery/tables/applications/posix/prometheus_metrics_utils.h [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+class PrometheusMetricsConfigParserPlugin : public ConfigParserPlugin {
+ public:
+  std::vector<std::string> keys() const override {
```
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics_utils.h]: gotcha

File: osquery/tables/applications/posix/prometheus_metrics.h (hunk 9 above)
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.h]: No, the capital "S" was to indicate seconds.

Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.h]: yup you're right

**Comment by allanliu** [comment]
Per my conversation with @PoppySeedPlehzr, I'm going to remove the class encapsulation of this table implementation altogether.   Given how simple this table is, this extra layer of abstraction really offers no benefits to offset the additional complexity.

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    }
+  }
+}
+
+QueryData& PrometheusMetrics::queryPrometheusTargets() {
+  scrapeTargets();
+  parseScrapeResults();
```
Comment (jedi22) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: This function that relies entirely on side-effects is hard to follow. Instead, try to structure it so that you pass things to a function and use what the function returns. This makes code easier to follow and debug. Could you pass scrapeResults to scrapeTargets as a parameter? Same with parseScrapeResults?

File: osquery/tables/applications/posix/prometheus_metrics.h (hunk 9 above)
Comment (jedi22) [inline osquery/tables/applications/posix/prometheus_metrics.h]: timeoutSeconds might be a better name

**Comment by allanliu** [comment]
Changes made :)

File: osquery/tables/applications/posix/prometheus_metrics.h [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+#include <vector>
+
+#include <osquery/tables.h>
+
+#include <boost/network/protocol/http/client.hpp>
+
+namespace osquery {
+namespace tables {
+
+const std::string kcol_target_name = "target_name";
```
Comment (jedi22) [inline osquery/tables/applications/posix/prometheus_metrics.h]: Global variables should have the form kColTargetName

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+  for (auto const& target : scrapeResults) {
+    std::stringstream ss(target.second.content);
+    std::string dest;
+    std::string ts(std::to_string(target.second.timestampMS.count()));
+
+    while (std::getline(ss, dest)) {
+      if (!dest.empty() && dest[0] != '#') {
+        std::vector<std::string> metric(osquery::split(dest, " "));
+
+        if (metric.size() > 1) {
```
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: yeah, if there's less than that number of elements, we don't have a full metric pair.  In the wild there should always be 2.  I used "> 1" just in case there's some unexpected character (i.e '\t' or something)

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+          rows.push_back(r);
+        }
+      }
+    }
+  }
+
+  return rows;
+}
+
+std::map<std::string, retData>& scrapeTargets(
```
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: good catch.  will fix

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+  std::map<std::string, retData> sr;
+  for (const auto& url : config.get_child("urls")) {
+    if (!url.first.empty()) {
+      return result;
+    }
+
+    sr[url.second.data()] = retData{};
+  }
+
+  return parseScrapeResults(scrapeTargets(sr), result);
```
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: You mean break out the scrapeTargets call from the parameter of parseScrapeResults?

File: osquery/tables/applications/posix/prometheus_metrics.h (hunk 16 above)
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.h]: So i'm bit confused, i checked with @PoppySeedPlehzr about the const names last night before changing.   I see in the osquery source, there's both snake cased and camel cased names...

File: osquery/tables/applications/posix/prometheus_metrics.cpp (hunk 19 above)
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: nvm, i see what you mean

File: osquery/tables/applications/posix/prometheus_metrics.h (hunk 16 above)
Comment (jedi22) [inline osquery/tables/applications/posix/prometheus_metrics.h]: Hmmm, I will check with @PoppySeedPlehzr

Comment (jedi22) [inline osquery/tables/applications/posix/prometheus_metrics.h]: Sorry if we weren't clear, he may have been talking about `k` prefixes but the standard should be camelcase and for constants, to start with a `k`.

Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.h]: Cool.  Thanks for the clarification! Will fix.

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+
+QueryData genPrometheusMetrics(QueryContext& context) {
+  QueryData result;
+  // get urls from config
+  auto parser = Config::getParser("prometheus_targets");
+  if (parser == nullptr || parser.get() == nullptr) {
+    return result;
+  }
+
+  const auto& config = parser->getData().get_child(kConfigParserRootKey);
```
Comment (jedi22) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: Can this throw? If it does, you'll need a try and catch or perhaps another function that contains a default value.

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+  if (config.count("urls") == 0) {
+    return result;
+  }
+
+  std::map<std::string, retData> sr;
+  for (const auto& url : config.get_child("urls")) {
```
Comment (jedi22) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: Same as above

Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: If this [line](https://github.com/digitalocean/osquery/blob/0ca1fc84bd6d52bd4126325da2f553439c011e6d/osquery/tables/applications/posix/prometheus_metrics.cpp#L81) evaluates to true, then the function returns, and the subject line of code would never hit.

File: osquery/tables/applications/posix/prometheus_metrics.cpp (hunk 20 above)
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: Yeah this does throw, but the get_child param is a const value for the config root key defined in prometheus_metrics_utils.h, which is also used by PrometheusMetricsConfigParserPlugin (in prometheus_metrics_utils.cpp).   In the wild, an exception should not be thrown by this call, and if it does shouldn't we consider this an exceptional event and the app should crash? If the prometheus_metrics configuration file is not supplied, this [line](https://github.com/digitalocean/osquery/blob/0ca1fc84bd6d52bd4126325da2f553439c011e6d/osquery/tables/applications/posix/prometheus_metrics.cpp#L81) evaluates to true and function returns an empty QueryData.

Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: Nah, if this does somehow magically throw in the wild we don't want the daemon to come to a grinding halt. For this, use a non-throwing call with a default value like 'get_child(kConfigParserRootKey,"")' and check the output.

File: osquery/tables/applications/posix/prometheus_metrics.cpp (hunk 21 above)
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: That's a good point. You're definitely right, but for consistency would you mind using the non throwing equivalent?

File: osquery/tables/applications/posix/prometheus_metrics.cpp (hunk 20 above)
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: In this case, would it better to try/catch this guy and log an error for the exception thrown if does occur?

File: osquery/tables/applications/posix/prometheus_metrics.cpp (hunk 21 above)
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: roger that.

**Comment by allanliu** [comment]
changes made :)

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+void scrapeTargets(std::map<std::string, retData>& scrapeResults) {
+  http::client client(
+      http::client::options().follow_redirects(true).timeout(1));
+
+  for (auto& target : scrapeResults) {
+    try {
+      http::client::request request(target.first);
+      http::client::response response(client.get(request));
+
+      target.second.timestampMS =
```
Comment (jedi22) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: Out of curiosity, why is this in milliseconds? Generally timestamps are in seconds and this will give it more consistency with other tables, is there a reason you want the extra precision?

Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: In the prometheus data model, metric sample timestamps are of millisecond precision ([link](https://prometheus.io/docs/concepts/data_model/)).

**Comment by theopolis** [comment]
Do you mind placing the config parser in `osquery/tables/applications/posix/prometheus_metrics_utils.cpp` alongside the other parsers in the config/parsers directory?

**Comment by allanliu** [comment]
Moved the config parser into the config/parsers directory.  Seeing that all the other config parsers are unit tested, I  will add tests for PrometheusMetricsConfigParserPlugin later and put it as a separate PR (? let me know if otherwise).

BTW, there's another config parser located in osquery/tables/other/yara_utils.cpp.

**Comment by jedi22** [comment]
ok to test

**Comment by theopolis** [comment]
ok to test

File: osquery/config/parsers/prometheus_targets.cpp [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <iostream>
+
+#include <osquery/config.h>
+#include <osquery/logger.h>
+
+#include <osquery/config/parsers/prometheus_targets.h>
```
Comment (PoppySeedPlehzr) [inline osquery/config/parsers/prometheus_targets.cpp]: nit: Use `"` for relative path to includes:
`#include "osquery/config/parsers/prometheus_targets.h"`

File: osquery/config/parsers/prometheus_targets.h [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ */
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include <osquery/config.h>
+
+namespace pt = boost::property_tree;
```
Comment (PoppySeedPlehzr) [inline osquery/config/parsers/prometheus_targets.h]: nit: This came up in a PR that I have up right now, we try to avoid dropping aliases in header files. Try to make use of the full path if you don't mind for ease of reading declaration specs.

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+ */
+#include <sstream>
+
+#include <osquery/config.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include <osquery/core/conversions.h>
+
+#include <osquery/config/parsers/prometheus_targets.h>
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: nit: quotes for relative includes

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+#include <osquery/tables/applications/posix/prometheus_metrics.h>
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: same.

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+void parseScrapeResults(const std::map<std::string, retData>& scrapeResults,
+                        QueryData& rows) {
+  for (auto const& target : scrapeResults) {
+    std::stringstream ss(target.second.content);
+    std::string dest;
+    std::string ts(std::to_string(target.second.timestampMS.count()));
+
+    while (std::getline(ss, dest)) {
+      if (!dest.empty() && dest[0] != '#') {
+        std::vector<std::string> metric(osquery::split(dest, " "));
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: I think you can use `auto metric = osquery::split(dest, " ");` here?

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 28]
```diff
@@ -0,0 +1,121 @@
...
+
+QueryData genPrometheusMetrics(QueryContext& context) {
+  QueryData result;
+  // get urls from config
+  auto parser = Config::getParser("prometheus_targets");
+  if (parser == nullptr || parser.get() == nullptr) {
+    return result;
+  }
+
+  // Add a specific value to the default property tree to differentiate it
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: nit: We tend to prefer the `/* */` style commenting for multi-line comments

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 29]
```diff
@@ -0,0 +1,121 @@
...
+  // config.
+  const auto& config = parser->getData().get_child(kConfigParserRootKey,
+                                                   pt::ptree("UNEXPECTED"));
+  if (config.get_value("") == "UNEXPECTED") {
+    LOG(WARNING) << "could not load prometheus_targets root key: "
+                 << kConfigParserRootKey;
+    return result;
+  }
+  if (config.count("urls") == 0) {
+    // Only warn the user if they supplied the config, but did not supply any
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: Same.

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 30]
```diff
@@ -0,0 +1,121 @@
...
+    // urls.
+    if (!config.empty()) {
+      LOG(WARNING) << "prometheus_targets configuration is missing field: urls";
+    }
+
+    return result;
+  }
+
+  std::map<std::string, retData> sr;
+  // Below should be unreachable if there were no urls child node, but we set
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: same.

File: osquery/tables/applications/posix/prometheus_metrics.h [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+#pragma once
+
+#include <chrono>
+#include <map>
+#include <string>
+#include <vector>
+
+#include <osquery/tables.h>
+
+#include <boost/network/protocol/http/client.hpp>
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.h]: nit: We try to group header includes from broadest to most specific, so throw this above the `osquery/tables` include you have above

File: osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp [hunk 32]
```diff
@@ -0,0 +1,416 @@
...
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+#include <gtest/gtest.h>
+
+#include <osquery/tables/applications/posix/prometheus_metrics.h>
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp]: nit: quotes for relative path includes.

File: osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp [hunk 33]
```diff
@@ -0,0 +1,416 @@
...
+    EXPECT_TRUE(expected[i][kColValue] == got[i][kColValue]);
+    EXPECT_TRUE(expected[i][kColTimeStamp] == got[i][kColTimeStamp]);
+  }
+}
+
+class PrometheusMetricsTest : public ::testing::Test {};
+
+TEST_F(PrometheusMetricsTest, no_targets) {
+  std::map<std::string, retData> sr;
+  // PrometheusMetrics pm(urls);
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp]: nit: Any reason to leave this in? If it's not being used let's remove it.

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 34]
```diff
@@ -0,0 +1,121 @@
...
+
+namespace osquery {
+namespace tables {
+
+void parseScrapeResults(const std::map<std::string, retData>& scrapeResults,
+                        QueryData& rows) {
+  for (auto const& target : scrapeResults) {
+    std::stringstream ss(target.second.content);
+    std::string dest;
+    std::string ts(std::to_string(target.second.timestampMS.count()));
```
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: As you're only using this to populate the row below, let's reduce it's scope and remove this line. Then down below you could have:
`r[kColTimeStamp] = BIGINT(target.second.timestampMS.count())`

File: osquery/tables/applications/posix/prometheus_metrics.cpp (hunk 27 above)
Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: I thought it would be clearer to leave the type when initializing the metric variable here...still change?

File: osquery/config/parsers/prometheus_targets.cpp (hunk 23 above)
Comment (allanliu) [inline osquery/config/parsers/prometheus_targets.cpp]: will fix all that have no others comments.

File: osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp (hunk 33 above)
Comment (allanliu) [inline osquery/tables/applications/posix/tests/prometheus_metrics_tests.cpp]: nope, just forgot to delete it.  will remove.

File: osquery/tables/applications/posix/prometheus_metrics.cpp (hunk 27 above)
Comment (PoppySeedPlehzr) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: Nah it's fine. We try to prefer 'auto' if at all possible, but if you'd like the clarity I don't think it's a huge deal.

Comment (allanliu) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: Ah.  Well then in that case, I'll change to auto.  Rather have it be consistent with the rest of the project :)

**Comment by allanliu** [comment]
Changes made.

**Comment by osqueryer** [comment]
:no_good: The commit 9aaf731060d1afaf6814ba357ce0061a23f8eb30 (Job results: [1033](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1033/)) failed the code audit or documentation test.

File: osquery/tables/applications/posix/prometheus_metrics.cpp [hunk 35]
```diff
@@ -0,0 +1,122 @@
...
+      http::client::request request(target.first);
+      http::client::response response(client.get(request));
+
+      target.second.timestampMS =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              std::chrono::system_clock::now().time_since_epoch());
+      target.second.content = static_cast<std::string>(body(response));
+
+    } catch (std::exception& e) {
+      LOG(ERROR) << "failed on scrape of target " << target.first << ": "
```
Comment (theopolis) [inline osquery/tables/applications/posix/prometheus_metrics.cpp]: nit, please capitalize the first letter in each log line.

File: osquery/tables/applications/posix/prometheus_metrics.h [hunk 36]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace osquery {
+namespace tables {
+
+const std::string kColTargetName = "target_name";
+const std::string kColMetric = "metric_name";
+const std::string kColValue = "metric_value";
+const std::string kColTimeStamp = "timestamp_ms";
+
+struct retData {
```
Comment (theopolis) [inline osquery/tables/applications/posix/prometheus_metrics.h]: Please name this something more-related to the table implementation, e.g.:
```cpp
struct PrometheusResponseData {
```

**Comment by osqueryer** [comment]
:-1: The commit 2f5d2d45cfa968b85f920c428d83b15d0d6d1a31 (Job results: [393](https://jenkins.osquery.io/job/osqueryPullRequestOSX/393/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:no_good: The commit 2f5d2d45cfa968b85f920c428d83b15d0d6d1a31 (Job results: [1094](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1094/)) failed the code audit or documentation test.

**Comment by allanliu** [comment]
Changes made.

**Comment by theopolis** [comment]
@poppyseedplehzr  or @jedi22 this looks good to me, I'll let either of you give it one more pass and feel free to merge depending on your review.
